### PR TITLE
fix potential panic on http server close

### PR DESCRIPTION
**Pull request by cpuguy83**

### CODE REVIEW

File: cmd/http.go [hunk 1]
```diff
@@ -57,43 +60,46 @@ func setupHTTPServer(ctx context.Context, cfg *apiServerConfig) (io.Closer, io.C
...
 	if cfg.MetricsAddr == "" {
 		log.G(ctx).Info("Pod metrics server not setup due to empty metrics address")
 	} else {
 		l, err := net.Listen("tcp", cfg.MetricsAddr)
 		if err != nil {
-			if l != nil {
-				podS.Close()
-			}
-			return nil, nil, errors.Wrap(err, "could not setup listenr for pod metrics http server")
+			cancel()
```
Comment (pires) [inline cmd/http.go]: Why call `cancel` here and return `cancel`? I'd expect the caller to call `cancel` if this function returns error != nil.

File: cmd/root.go [hunk 2]
```diff
@@ -108,13 +108,11 @@ This allows users to schedule kubernetes workloads on nodes that aren't running
 			rootContextCancel()
 		}()
 
-		c1, c2, err := setupHTTPServer(rootContext, apiConfig)
+		cancelHTTP, err := setupHTTPServer(rootContext, apiConfig)
 		if err != nil {
 			log.G(rootContext).Fatal(err)
```
Comment (pires) [inline cmd/root.go]: Call `cancelHTTP`?

Comment (cpuguy83) [reply cmd/root.go]: I think we should just have cancel be nil when there is an error and call cancel in the setup function. Updating this.

File: cmd/http.go [hunk 3]
```diff
@@ -44,11 +44,19 @@ func loadTLSConfig(certPath, keyPath string) (*tls.Config, error) {
...
+func setupHTTPServer(ctx context.Context, cfg *apiServerConfig) (cancel func(), retErr error) {
+	var closers []io.Closer
+	cancel = func() {
+		for _, c := range closers {
+			c.Close()
+		}
+	}
+	defer func() {
+		if retErr != nil {
+			cancel()
```
Comment (pires) [inline cmd/http.go]: There's no more the need to define `cancel` func then. Can we just move the `for` loop above here?

Comment (cpuguy83) [reply cmd/http.go]: Sorry, we still need cancel because it is returned to the caller and used here.

Comment (pires) [reply cmd/http.go]: I get it now. If it errors, cancel happens immediately. If not, the caller will cancel later.

**Pull request merged**